temporalio/common.py

```diff
@@ -1063,6 +1065,65 @@ def from_canonical_string(canonical: str) -> WorkerDeploymentVersion:
             )
         return WorkerDeploymentVersion(parts[0], parts[1])
 
+    def _to_proto(self) -> temporalio.api.deployment.v1.WorkerDeploymentVersion:
+        """Convert to proto representation."""
+        return temporalio.api.deployment.v1.WorkerDeploymentVersion(
+            build_id=self.build_id,
+            deployment_name=self.deployment_name,
+        )
+
+
+class VersioningOverride(ABC):
+    """Represents the override of a worker's versioning behavior for a workflow execution.
+
+    .. warning::
+        Experimental API.
+    """
+
+    @abstractmethod
+    def _to_proto(self) -> temporalio.api.workflow.v1.VersioningOverride:
+        """Convert to proto representation."""
+        pass
+
+
+@dataclass(frozen=True)
+class PinnedVersioningOverride(VersioningOverride):
+    """Workflow will be pinned to a specific deployment version.
+
+    .. warning::
+        Experimental API.
+    """
+
+    version: WorkerDeploymentVersion
+
+    def _to_proto(self) -> temporalio.api.workflow.v1.VersioningOverride:
+        """Convert to proto representation."""
+        # Include deprecated fields for backward compatibility
+        return temporalio.api.workflow.v1.VersioningOverride(
+            behavior=temporalio.api.enums.v1.VersioningBehavior.VERSIONING_BEHAVIOR_PINNED,
+            pinned_version=self.version.to_canonical_string(),
+            pinned=temporalio.api.workflow.v1.VersioningOverride.PinnedOverride(
+                version=self.version._to_proto(),
+                behavior=temporalio.api.workflow.v1.VersioningOverride.PinnedOverrideBehavior.PINNED_OVERRIDE_BEHAVIOR_PINNED,
+            ),
+        )
+
+
+class AutoUpgradeVersioningOverride(VersioningOverride):
```

Comment:
Still probably worth a frozen dataclass here so you get the other goodies (hash, str, etc)